File: Linked List/Medium/Finding loop in linkedlist/Findingloop.py
class Node:
    def __init__(self, val):
        self.data = val
        self.next = None

class SLL:
    def __init__(self):
        self.head = None

    def insert_begin(self, val):
        # Insert a new node at the beginning of the linked list
        new_node = Node(val)
        new_node.next = self.head
        self.head = new_node

    def insert_end(self, val):
        # Insert a new node at the end of the linked list
        new_node = Node(val)
        if self.head is None:
            self.head = new_node
        else:
            temp = self.head
            while temp.next:
                temp = temp.next
            temp.next = new_node

    def insert_pos(self, val, pos):
        # Insert a new node at a specified position in the linked list
        if pos < 1:
            print("Invalid position for insertion.")
            return

        count = 0
        temp = self.head
        while temp:
            count += 1
            if count == pos - 1:
                break
            temp = temp.next

        if temp is None:
            print("Invalid position for insertion.")
            return

        new_node = Node(val)
        new_node.next = temp.next
        temp.next = new_node

    def create_loop(self, position, by_value=True):
        # Create a loop in the linked list either by value or position
        count = 0
        temp = self.head
        while temp.next:
            count += 1
            if (by_value and count == position) or (not by_value and count == position - 1):
                loop_node = temp.next
                break
            temp = temp.next

        temp.next = loop_node

    def detect_loop(self):
        # Detect a loop in the linked list using Floyd's Tortoise and Hare algorithm
        slow = self.head
        fast = self.head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                print("Loop detected in the linked list.")
                return slow  # Return the node where the loop starts

        print("No loop found in the linked list.")
        return None

    def remove_loop(self):
        # Remove the loop in the linked list using Floyd's Tortoise and Hare algorithm
        loop_start = self.detect_loop()
        if loop_start:
            slow = self.head
            fast = loop_start

            while slow.next != fast.next:
                slow = slow.next
                fast = fast.next

            fast.next = None

    def display(self):
        # Display the elements of the linked list
        temp = self.head
        while temp:
            print(temp.data, end=" -> ")
            temp = temp.next
        print("None")


# Test drive code with user input:
sll = SLL()

print("------------INSERTION OPERATIONS-------------")
n = int(input("Enter the number of elements to insert at the beginning: "))
for _ in range(n):
    val = int(input("Enter the value to insert at the beginning: "))
    sll.insert_begin(val)
sll.display()

n = int(input("Enter the number of elements to insert at the end: "))
for _ in range(n):
    val = int(input("Enter the value to insert at the end: "))
    sll.insert_end(val)
sll.display()

val = int(input("Enter the value to insert: "))
pos = int(input("Enter the position to insert at: "))
sll.insert_pos(val, pos)
sll.display()

print("------------LOOP CREATION-------------")
pos = int(input("Enter the position for loop creation: "))
sll.create_loop(pos)
sll.display()

print("------------LOOP DETECTION-------------")
sll.detect_loop()

print("------------LOOP REMOVAL-------------")
sll.remove_loop()
sll.display()

File: Linked List/Medium/Finding loop in linkedlist/FindingLoop.cpp
#include <iostream>
using namespace std;

/* Node class definition */
class VSDnode {
public:
    int data;
    VSDnode* next;

    VSDnode(int val) {
        data = val;
        next = NULL;
    }
};

/* Function to add a node at the head of the linked list */
void add_head(VSDnode* &head, int val) {
    VSDnode* n = new VSDnode(val);
    n->next = head;
    head = n;
}

/* Function to add a node at the tail of the linked list */
void add_tail(VSDnode* &head, int val) {
    VSDnode* n = new VSDnode(val);
    if (head == NULL) {
        head = n;
        return;
    }
    VSDnode* temp = head;
    while (temp->next != NULL) {
        temp = temp->next;
    }
    temp->next = n;
}

/* Function to create a cycle in the linked list at a specified position */
void create_cycle_pos(VSDnode* &head, int pos) {
    VSDnode* temp = head;
    int len = 1;
    VSDnode* ptr;
    while (temp->next != NULL) {
        if (len == pos) {
            ptr = temp;
        }
        temp = temp->next;
        len++;
    }
    temp->next = ptr;
}

/* Function to detect the presence of a cycle in the linked list using Floyd's Tortoise and Hare algorithm */
bool detect_cycle(VSDnode* &head) {
    VSDnode* slow = head;
    VSDnode* fast = head;
    while (fast != NULL && fast->next != NULL) {
        slow = slow->next;
        fast = fast->next->next;
        if (fast == slow) {
            return true;  /* Cycle present in the linked list */
        }
    }
    return false;  /* No cycle present in the linked list */
}

/* Function to remove the cycle in the linked list */
void remove_cycle(VSDnode* &head) {
    VSDnode* slow = head;
    VSDnode* fast = head;

    do {
        slow = slow->next;
        fast = fast->next->next;
    } while (slow != fast);

    fast = head;
    while (slow->next != fast->next) {
        slow = slow->next;
        fast = fast->next;
    }
    slow->next = NULL;
}

/* Function to display the linked list */
void display(VSDnode* head) {
    VSDnode* temp = head;
    while (temp != NULL) {
        cout << temp->data << "->";
        temp = temp->next;
    }
    cout << "NULL" << endl;
}

int main() {
    VSDnode* head = NULL;
    int n;

    cout << "Enter the number of elements in the linked list: ";
    cin >> n;

    cout << "Enter the elements of the linked list:" << endl;
    for (int i = 0; i < n; i++) {
        int element;
        cin >> element;
        add_tail(head, element);
    }

    cout << "Do you want to create a cycle in the linked list? (1 for Yes, 0 for No): ";
    int createCycle;
    cin >> createCycle;

    if (createCycle) {
        cout << "Enter the position to create a cycle: ";
        int pos;
        cin >> pos;
        create_cycle_pos(head, pos);
    }

    cout << "Linked List: ";
    display(head);

    /* Detecting and displaying whether a cycle is present */
    if (detect_cycle(head)) {
        cout << "Cycle present in the linked list" << endl;
        remove_cycle(head);
        cout << "Linked List after removing the cycle: ";
        display(head);
    } else {
        cout << "No cycle present in the linked list" << endl;
    }

    return 0;
}
